src/execution/operator/join/physical_hash_join.cpp

```diff
@@ -68,45 +87,126 @@ PhysicalHashJoin::PhysicalHashJoin(PhysicalPlan &physical_plan, LogicalOperator
+				if (it == build_columns_in_conditions.end()) {
+					idx_t layout_pos = condition_types.size() + payload_columns.col_idxs.size();
+					build_input_to_layout[rhs_idx_with_offset] = layout_pos;
+
+					payload_columns.col_idxs.push_back(rhs_idx);
+					payload_columns.col_types.push_back(rhs_input_types[rhs_idx]);
+				}
+			}
+		}
+
+		build_input_to_layout_map = std::move(build_input_to_layout);
 		return;
 	}
 
-	auto &rhs_input_types = children[1].get().GetTypes();
-
-	// Create a projection map for the RHS (if it was empty), for convenience
+	// for other join types
 	auto right_projection_map_copy = right_projection_map;
 	if (right_projection_map_copy.empty()) {
-		right_projection_map_copy.reserve(rhs_input_types.size());
 		for (idx_t i = 0; i < rhs_input_types.size(); i++) {
 			right_projection_map_copy.emplace_back(i);
 		}
 	}
 
-	// Now fill payload expressions/types and RHS columns/types
+	for (auto rhs_idx_with_offset : predicate_build_cols) {
+		idx_t rhs_idx = rhs_idx_with_offset - lhs_input_types.size();
+
+		// skip if it's a join condition
+		auto it = build_columns_in_conditions.find(rhs_idx);
+		if (it == build_columns_in_conditions.end()) {
+			idx_t layout_pos = condition_types.size() + payload_columns.col_idxs.size();
+			build_input_to_layout[rhs_idx_with_offset] = layout_pos;
+
+			payload_columns.col_idxs.push_back(rhs_idx);
+			payload_columns.col_types.push_back(rhs_input_types[rhs_idx]);
+		}
+	}
+
 	for (auto &rhs_col : right_projection_map_copy) {
 		auto &rhs_col_type = rhs_input_types[rhs_col];
+		idx_t rhs_col_with_offset = lhs_input_types.size() + rhs_col;
 
+		// check if it's a join condition
 		auto it = build_columns_in_conditions.find(rhs_col);
-		if (it == build_columns_in_conditions.end()) {
-			// This rhs column is not a join key
-			payload_columns.col_idxs.push_back(rhs_col);
-			payload_columns.col_types.push_back(rhs_col_type);
-			rhs_output_columns.col_idxs.push_back(condition_types.size() + payload_columns.col_types.size() - 1);
-		} else {
-			// This rhs column is a join key
+
+		if (it != build_columns_in_conditions.end()) {
+			// it's in join conditions - use condition position
 			rhs_output_columns.col_idxs.push_back(it->second);
+		} else {
+			// check if already in payload (from predicate)
+			auto layout_it = build_input_to_layout.find(rhs_col_with_offset);
+
+			if (layout_it != build_input_to_layout.end()) {
+				// already in payload (from predicate) - reuse position
+				rhs_output_columns.col_idxs.push_back(layout_it->second);
+			} else {
+				// new column - add to payload
+				idx_t layout_pos = condition_types.size() + payload_columns.col_idxs.size();
+				payload_columns.col_idxs.push_back(rhs_col);
+				payload_columns.col_types.push_back(rhs_col_type);
+				rhs_output_columns.col_idxs.push_back(layout_pos);
+			}
 		}
+
 		rhs_output_columns.col_types.push_back(rhs_col_type);
 	}
+
+	build_input_to_layout_map = std::move(build_input_to_layout);
```

Comment:
The constructor of `PhysicalHashJoin` is now >150 lines. Would it be possible to move the code related to the residual predicate into a separate initialization function?